src/base/Join.hh

```diff
@@ -0,0 +1,79 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file Join.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "detail/Joined.hh"
+
+namespace celeritas
```

Comment:
Should we have an "internal" sub-namespace to indicate this type of utility routine that are not (fully) part of the API ... humm actually maybe the question is why is ```Joined``` in detail and not ```join```

Reply:
So here I don't expect any of the rest of the codebase to use the `Joined` class -- except for the `join` function. To keep the header cleaner and hide the implementation the implementation (a class called `Joined`) is in the detail directory.

Reply:
I see.  I missed the the part "the header file is in src", which indeed is also indicative of a private/detail header.  Since it is nowadays (ide and githubs) easy to miss the file location, I ponder if we should not only rely on that location but also the namespace (i.e. have this file declare its entities inside celeritas::detail and/or celertias::internal and/or celeritas::private).

Reply:
@pcanal I think that's what I'm doing now: classes/functions should be in the `celeritas::detail` namespace if and only if they're inside a `detail/` subdirectory. Am I missing something?

Reply:
@sethrj humm ... well ... I managed to get myself even more confused by mentally using the wrong project's file location convention  (i.e. celeritas has src/[package]/*.[cc|hh] where those header are public some other project has both [package]/inc/*.hh and [package/src/*.[cc|hh] where the later header file are 'private']).

So, if I understand correctly
* "base/Join.hh" is a public header
* it declares ```celeritas::join```

So given that celeritas has a single private-ish namesped (`detail`), I still believe/feel that the join function should be there.

Or in other word, is the current ```celeritas::join``` intended to be used (and relied upon) by user code or is it an implementation detail/helper for celeritas' own code?

Reply:
`celeritas::join` is indeed intended to be "public", one of the utility functions available to apps etc.

Since we're still working basically as a "standalone app" type repository, I've been somewhat glib about separating public and private stuff. To first order we will direct cmake to install all `.hh` files except those in `detail` subdirectories. If/when we need to transition to a full library we'll probably need to refine that to account for detail headers that get included by public headers.

It's also conceivable that as Celeritas matures we'll export most of the code in `base` to a separate utility library for additional use in Shift (if we ever want to refactor it to use the dual host/device code strategies in Celeritas). In that case `join` will replace the SCALE utility whence it was derived :)